src/Engine/Combat/CombatEngine.java

```diff
@@ -25,10 +34,74 @@ public static BattleSummary calculateBattleResults( Unit attacker, Unit defender
+    WeaponModel attackerWeapon = attacker.chooseWeapon(defender.model, battleRange, attackerMoved);
+    WeaponModel defenderWeapon = null;
+    // Only attacks at point-blank range can be countered
+    if( 1 == battleRange )
+      defenderWeapon = defender.chooseWeapon(attacker.model, battleRange, false);
+
+    CombatContext context = new CombatContext(map, attacker, attackerWeapon, defender, defenderWeapon, battleRange, attackerX, attackerY);
+    
+    // If the attacker and defender get swapped, we want to still give a coherent BattleSummary.
+    Map<Unit, Entry<WeaponModel,Double>> unitDamageMap = new HashMap<Unit, Entry<WeaponModel,Double>>();
+    unitDamageMap.put(attacker, new AbstractMap.SimpleEntry<WeaponModel,Double>(attackerWeapon, 0.0));
+    unitDamageMap.put(defender, new AbstractMap.SimpleEntry<WeaponModel,Double>(defenderWeapon, 0.0));
 
-    return params.calculateBattleResults();
+    // From here on in, use context variables only
+
+    double attackerHPLoss = 0;
+
+    // Set up our scenario.
+    BattleParams attackInstance = StrikeParams.getAttack(context);
+
+    double defenderHPLoss = attackInstance.calculateDamage();
+    unitDamageMap.put(context.attacker, new AbstractMap.SimpleEntry<WeaponModel,Double>(context.attackerWeapon, defenderHPLoss));
+    if( !isSim && defenderHPLoss > context.defender.getPreciseHP() ) defenderHPLoss = context.defender.getPreciseHP();
+
+    // If the unit can counter, and wasn't killed in the initial volley, calculate return damage.
+    if( context.canCounter && (context.defender.getPreciseHP() > defenderHPLoss) )
+    {
+      // New battle instance with defender counter-attacking.
+      double counterHP = Math.ceil(context.defender.getPreciseHP() - defenderHPLoss); // Account for the first attack's damage to the now-attacker.
+      BattleParams defendInstance = StrikeParams.getCounterAttack(context, counterHP);
+
+      attackerHPLoss = defendInstance.calculateDamage();
+      unitDamageMap.put(context.defender, new AbstractMap.SimpleEntry<WeaponModel,Double>(context.defenderWeapon, attackerHPLoss));
+      if( !isSim && attackerHPLoss > context.attacker.getPreciseHP() ) attackerHPLoss = context.attacker.getPreciseHP();
+    }
+    
+    // Calculations complete. We are setting up our BattleSummary, so go back to using known-accurate info
+    return new BattleSummary(attacker, unitDamageMap.get(attacker).getKey(),
+                             defender, unitDamageMap.get(defender).getKey(),
+                             map.getEnvironment(attackerX, attackerY).terrainType,
+                             map.getEnvironment(defenderX, defenderY).terrainType,
+                             unitDamageMap.get(defender).getValue(), unitDamageMap.get(attacker).getValue());
   }
 
+  public static BattleParams buildSimpleAttack( Unit attacker, int battleRange, Unit defender, GameMap map, int terrainStars, boolean attackerMoved )
```

Comment:
Is the attack "simple" just because it doesn't account for the counterattack?
I'd almost prefer to call this `calculateAttackDamage` or something, and just have it return just the pain rather than the whole `BattleParams` (since nobody uses that at this point anyway).

Reply:
You can push changes whenever. I've looked over the whole first batch of changes.